evm/src/cpu/kernel/asm/memory/syscalls.asm

```diff
@@ -92,12 +92,48 @@ calldataload_large_offset:
     // stack: offset, total_size, kexit_info, dest_offset, offset, size
     GT %jumpi(wcopy_large_offset)
 
+    PUSH $segment
+    %mload_context_metadata($context_metadata_size)
+    // stack: total_size, segment, kexit_info, dest_offset, offset, size
+    DUP6 DUP6 ADD
+    // stack: offset + size, total_size, segment, kexit_info, dest_offset, offset, size
+    LT %jumpi(wcopy_within_bounds)
+
+    %mload_context_metadata($context_metadata_size)
+    // stack: total_size, segment, kexit_info, dest_offset, offset, size
+    DUP6 DUP6 ADD
+    // stack: offset + size, total_size, segment, kexit_info, dest_offset, offset, size
+    SUB // extra_size = offset + size - total_size
+    // stack: extra_size, segment, kexit_info, dest_offset, offset, size
+    DUP1 DUP7 SUB
+    // stack: copy_size = size - extra_size, extra_size, segment, kexit_info, dest_offset, offset, size
+
+    // Compute the new dest_offset after actual copies, at which we will start padding with zeroes.
+    DUP1 DUP6 ADD
+    // stack: new_dest_offset, copy_size, extra_size, segment, kexit_info, dest_offset, offset, size
+
     GET_CONTEXT
-    %stack (context, kexit_info, dest_offset, offset, size) ->
-        (context, @SEGMENT_MAIN_MEMORY, dest_offset, context, $segment, offset, size, wcopy_after, kexit_info)
+    %stack (context, new_dest_offset, copy_size, extra_size, segment, kexit_info, dest_offset, offset, size) ->
+        (context, @SEGMENT_MAIN_MEMORY, dest_offset, context, segment, offset, copy_size, wcopy_over_range, new_dest_offset, extra_size, kexit_info)
     %jump(memcpy_bytes)
 %endmacro
 
+wcopy_within_bounds:
+    // stack: segment, kexit_info, dest_offset, offset, size
+    GET_CONTEXT
+    %stack (context, segment, kexit_info, dest_offset, offset, size) ->
+        (context, @SEGMENT_MAIN_MEMORY, dest_offset, context, segment, offset, size, wcopy_after, kexit_info)
+    %jump(memcpy_bytes)
+
+
+// Same as wcopy_large_offset, but without `offset` in the stack.
+wcopy_over_range:
+    // stack: dest_offset, size, kexit_info
+    GET_CONTEXT
+    %stack (context, dest_offset, size, kexit_info) ->
+        (context, @SEGMENT_MAIN_MEMORY, dest_offset, size, wcopy_after, kexit_info)
+    %jump(memset)
```

Comment:
Not sure these changes are necessary since these segments have zeros outside the bounds.

Reply:
Actually I'm not sure we can remove it. Say a contract calls CODECOPY twice with an overlapping destination. The YP states that we should fill with zeros after reaching the end of the code, but if we do nothing we may hit non zero values from the previous copy ?

Reply:
Ah right, I forgot that CODECOPY also uses `wcopy`. What I said holds for calldata and returndata.
So yeah maybe it's fine to keep your changes for all of them to avoid duplicating this logic.

Reply:
Yeah I was going to suggest something along those lines. The extra overflow check from RETURNDATACOPY can't hurt CALLDATACOPY, and at least we'd have something cleaner and a tad more efficient.

Reply:
Actually it may take a bit more time to deal with it as CODECOPY / EXTCODECOPY have slight differences. Do you mind if we merge as is and I deal with it later?

Reply:
Yes sounds good 👍